_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,556 @@
+query to the database.
+
+```java
+public List<User> getUserByFirstWithProjection(String first, 
+     int max, 
+     int page){
+
+  FullTextEntityManager fullTextEntityManager = 
+    Search.getFullTextEntityManager(entityManager);
+  QueryBuilder qb = fullTextEntityManager
+    .getSearchFactory()
+    .buildQueryBuilder()
+    .forEntity(User.class)
+    .get();
+  
+  Query similarToUser = qb.keyword().fuzzy()
+    .withEditDistanceUpTo(2)
+    .onField("first")
+    .matching(first)
+    .createQuery();
+  
+  Query finalQuery = qb.bool()
+    .must(similarToUser)
+    .createQuery();
+  
+  FullTextQuery fullTextQuery = 
+    fullTextEntityManager.createFullTextQuery(
+      finalQuery,
+      User.class);
+
+  fullTextQuery.setProjection(
+    FullTextQuery.ID,
+    "first",
+    "last",
+    "middle",
+    "age");
+  fullTextQuery.setSort(qb.sort()
+    .byField("age")
+    .desc()
+    .andByScore()
+    .createSort());
+  fullTextQuery.setMaxResults(max);
+  fullTextQuery.setFirstResult(page);
+  
+  return getUserList(fullTextQuery.getResultList());
+}
+
+private List<User> getUserList(List<Object[]> resultList) {
+  List<User> users = new ArrayList<>();
+  for (Object[] objects : resultList) {
+      User user = new User();
+      user.setId((String) objects[0]);
+      user.setFirst((String) objects[1]);
+      user.setLast((String) objects[2]);
+      user.setMiddle((String) objects[3]);
+      user.setAge((Integer) objects[4]);
+      users.add(user);
+  }
+  return users;
+}
+```
+
+To use projection we need to pass the list of fields that we want in output in the `setProjection` method. 
+
+Now when we fetch results hibernate will return a list of object arrays. Which to be honest is a bit incontinent. Apart from
+fields, we can also fetch metadata such as id with `FullTextQuery.ID` or even score with `FullTextQuery.SCORE`.
+
+### Pagination
+```java
+FullTextQuery fullTextQuery = 
+   fullTextEntityManager.createFullTextQuery(
+     finalQuery,
+     User.class);
+//...
+fullTextQuery.setSort(qb.sort()
+   .byField("age")
+   .desc()
+   .andByScore()
+   .createSort());
+fullTextQuery.setMaxResults(max);
+fullTextQuery.setFirstResult(page);
+```
+Lastly, let's talk about pagination and sorting as we don't want to fetch millions of records that we have stored in our 
+Elasticsearch indexes.
+
+To perform pagination we need two things, the number of results we want per page and page offset (Page number, if I put plainly).
+Prior we can pass to `FullTextQuery` objects `setMaxResult` method and later in `setFirstResult`. Then query will return results 
+accordingly.
+
+Query DSL also provides us a way to define sort field and order using `sort()`. We can also perform sort operation on multiple fields by 
+chaining with `andByField()`.
+
+That's it! I mean this is not everything, but I believe this is enough to get you started. For further reading you can 
```

Comment:
This and below shouldn't be part of the Pagination section. You can call it "Further Reading" or something similar.